### PR TITLE
Modelcatalog

### PR DESCRIPTION
**Pull request by rupeshsahuoc**
https://github.com/oneconvergence/gpuaas/issues/7240

New function :- 
list_published_models(self, user) 
get_model_catalog(self, user, model)

### CODE REVIEW

**Comment by ajayanto** [comment]
You have to add support for   /users/{user}/publishedmodels, api

File: dkube/sdk/api.py [hunk 1]
```diff
@@ -1867,31 +1883,45 @@ def get_modelcatalog_item(self, user, modelcatalog=None, model=None, version=Non
...
         if version is None:
             return "Model Version must be provided"
-        if modelcatalog:
-            mc = self.modelcatalog(user)
-            for item in mc:
-                if item['name'] == modelcatalog:
-                    for iversion in item['versions']:
-                        if iversion['model']['version'] == version:
-                            return iversion
+        if pversion.parse(dkubever) < pversion.parse("2.2.1.13"):
```
Comment (ajayanto) [inline dkube/sdk/api.py:1892]: version should be < 2.3.0.0

File: dkube/sdk/api.py [hunk 2]
```diff
@@ -1867,31 +1883,45 @@ def get_modelcatalog_item(self, user, modelcatalog=None, model=None, version=Non
...
+                        for iversion in item['versions']:
+                            if iversion['model']['version'] == version:
+                                return iversion
 
-            raise Exception(
-                '{}.{} not found in model catalog'.format(model, version))
+                raise Exception(
+                    '{}.{} not found in model catalog'.format(model, version))
+            else:
+                mc = self.modelcatalog(user)
```
Comment (ajayanto) [inline dkube/sdk/api.py:1904]: call modelcatalog_model and return the result

File: dkube/sdk/api.py (hunk 1 above)
Comment (rupeshsahuoc) [reply dkube/sdk/api.py]: addressed

**Comment by hemanthravi** [comment]
@deepio-oc 
Can you review this PR

**Comment by deepio-oc** [comment]
@rupeshsahuoc list down the new APIs and how user is expected to call these APIs

**Comment by deepio-oc** [comment]
instead of making new APIs, lets try to reuse existing APIs where we can. Too many apis for same feature would be confusing to user.

1.instead of new get_modelcatalog, keep existing get_modelcatalog_item. in 2.3, you can ignore modelcatalog name argument. Update the function document

2. instead of published_models, use existing list_models API and enabled published=False argument.

@hemanthravi , what is your opinion on point 1 above? @rupeshsahuoc said that is the way he had earlier but changed as per @ajayanto  comment.

**Comment by ajayanto** [comment]
@deepio-oc  
1.the old get_modelcatalog_item api is getting all model catalog and filtering it.  This APi has multiple db calls. Inorder to prevent that we added new api.
2. The Api is entirely different. We have a published model list maintained at the backend. This new api is using that to fetch model details. 
The filter over a list using  list_models is a very costly operation, we have to fetch every model from the DB and do the filter, Here also we have to make too many unnecessary db calls in the backend. Due to that list model api is will become very slow and it will not return proper results in case of paging enabled

**Comment by deepio-oc** [comment]
@ajayanto 
can we keep the same API in SDK? inside the SDK implementation, we can call different APIs based on dkube version.

We dont need to map SDK API with the REST API. SDK API are customer facing so those should be driven by how customer should use than how we have implemented in the backend. 


> @deepio-oc
> 1.the old get_modelcatalog_item api is getting all model catalog and filtering it. This APi has multiple db calls. Inorder to prevent that we added new api.
> 2. The Api is entirely different. We have a published model list maintained at the backend. This new api is using that to fetch model details.
> The filter over a list using list_models is a very costly operation, we have to fetch every model from the DB and do the filter, Here also we have to make too many unnecessary db calls in the backend. Due to that list model api is will become very slow and it will not return proper results in case of paging enabled
> 
> These two new APIs are added for UI to reduce number of backend calls incase of published models

**Comment by ajayanto** [comment]
Yes @deepio-oc, we can do that.

**Comment by rupeshsahuoc** [comment]
@deepio-oc @ajayanto 
The get_modelcatalog_item() is used internally by create_model_deployment(). If the response structure of get_modelcatalog_item will change in new version, then there will be issue in create_model_deployment()

File: dkube/sdk/api.py [hunk 3]
```diff
@@ -1244,7 +1244,7 @@ def get_model(self, user, name):
 
         return super().get_repo('model', user, name)
 
-    def list_models(self, user, shared=False, filters='*'):
+    def list_models(self, user, shared=False, published=False, filters='*'):
```
Comment (mak-454) [inline dkube/sdk/api.py:1247]: the combination shared=True and published=False isnt correct right.. can we do a separate func list_published_models()

File: dkube/sdk/api.py [hunk 4]
```diff
@@ -1845,6 +1850,22 @@ def modelcatalog(self, user):
         """
         return super().modelcatalog(user)
 
+    def modelcatalog_model(self, user, model):
```
Comment (mak-454) [inline dkube/sdk/api.py:1853]: name doesnt seem ok.. can you rename to get_model_catalog

**Comment by mak-454** [comment]
@deepio-oc @ajayanto please wait till my review.. i just started
@hemanthravi

File: dkube/sdk/api.py [hunk 5]
```diff
@@ -1867,30 +1888,76 @@ def get_modelcatalog_item(self, user, modelcatalog=None, model=None, version=Non
                     Version of the model
 
         """
+        dkubever = self.dkubeinfo['version']
+        # ModelCatalog API is changed in 2.3.0.0 dkube version
         if modelcatalog is None and model is None:
             return "either model catalog name or model name should be provided"
         if version is None:
```
Comment (mak-454) [inline dkube/sdk/api.py:1895]: if version is mandatory argument why make it kwargs ? can you make it positional

File: dkube/sdk/api.py [hunk 6]
```diff
@@ -1845,6 +1850,22 @@ def modelcatalog(self, user):
...
+            *Inputs*
+
+                user
+                    Name of the user
+                model
+                    Name of the model.
+        """
+        return super().modelcatalog_model(user, model)
+
     def get_modelcatalog_item(self, user, modelcatalog=None, model=None, version=None):
```
Comment (mak-454) [inline dkube/sdk/api.py:1869]: what is modelcatalog_item ? I mean how would user know ? in this file we do not need to map to each and every BE API

File: dkube/sdk/api.py [hunk 7]
```diff
@@ -1867,30 +1888,76 @@ def get_modelcatalog_item(self, user, modelcatalog=None, model=None, version=Non
...
+                    if item['model']['name'] == model:
+                        for iversion in item['versions']:
+                            if iversion['model']['version'] == version:
+                                return iversion
+
+                raise Exception(
+                    '{}.{} not found in model catalog'.format(model, version))
+
+
+    def get_modelcatalog(self, user, model, version=None):
```
Comment (mak-454) [inline dkube/sdk/api.py:1930]: should this be list_modelcatalog ?

File: dkube/sdk/internal/api_base.py [hunk 8]
```diff
@@ -328,6 +328,14 @@ def modelcatalog(self, user):
         response = api.get_model_catalog(user)
         return response.to_dict()['data']
 
+    def modelcatalog_model(self, user, model):
```
Comment (mak-454) [inline dkube/sdk/internal/api_base.py:331]: get_modelcatalog ?

File: dkube/sdk/api.py (hunk 3 above)
Comment (rupeshsahuoc) [reply dkube/sdk/api.py]: addressed

File: dkube/sdk/api.py (hunk 4 above)
Comment (rupeshsahuoc) [reply dkube/sdk/api.py]: addressed

File: dkube/sdk/api.py (hunk 5 above)
Comment (rupeshsahuoc) [reply dkube/sdk/api.py:1901]: @mak-454  Can I make it get_modelcatalog_item(self, user, version, modelcatalog=None, model=None) ?
From now on wards order will matter

File: dkube/sdk/api.py (hunk 6 above)
Comment (rupeshsahuoc) [reply dkube/sdk/api.py:1875]: This function we used for get modelcatalog of a  specified version. This function is old

File: dkube/sdk/api.py (hunk 7 above)
Comment (rupeshsahuoc) [reply dkube/sdk/api.py]: get_model_catalog() will return the list of that model. So I removed this

File: dkube/sdk/internal/api_base.py [hunk 9]
```diff
@@ -328,6 +328,14 @@ def modelcatalog(self, user):
...
+        response = self._api.get_one_modelcatalog_model(user, model)
+        return response.to_dict()['data']
+
+    def published_models(self, user):
```
[the remark this answers is not in the document]
Comment (rupeshsahuoc) [reply dkube/sdk/internal/api_base.py]: addressed

File: dkube/sdk/internal/api_base.py (hunk 8 above)
Comment (rupeshsahuoc) [reply dkube/sdk/internal/api_base.py]: addressed

**Comment by hemanthravi** [comment]
@mak-454 @deepio-oc 
Can one of you merge this. This will address #7435.

@deepio-oc 
Automation test needs to be changes for this.

File: dkube/sdk/api.py (hunk 3 above)
Comment (deepio-oc) [reply dkube/sdk/api.py]: what is the issue with shared=True and Published=False ? If user wants to list the models which are shared by other users but not published, that's fine right?

**Comment by rupeshsahuoc** [comment]
Moved to PR #66

**Pull request closed without merging**